lib/instrumentation/modules/redis.js

```diff
@@ -6,11 +6,25 @@ var shimmer = require('../shimmer')
 var { getDBDestination } = require('../context')
 
 module.exports = function (redis, agent, { version, enabled }) {
-  if (!semver.satisfies(version, '^3.0.0')) {
+  if (!semver.satisfies(version, '>=2.0.0 <4.0.0')) {
     agent.logger.debug('redis version %s not supported - aborting...', version)
     return redis
   }
 
+  // The undocumented field on a RedisClient instance on which connection
+  // options are stored has changed a few times.
+  let connOptsFromRedisClient
+  if (semver.gte(version, '2.4.0')) {
+    connOptsFromRedisClient = (rc) => rc.connection_options
+  } else if (semver.gte(version, '2.3.0')) {
+    connOptsFromRedisClient = (rc) => rc.connection_option
+  } else if (semver.gte(version, '0.12.0')) {
+    connOptsFromRedisClient = (rc) => rc.connectionOption
+  } else {
+    // *Maybe* available on `client.{host|port}`, but not verified.
+    connOptsFromRedisClient = () => {}
+  }
+
```

Comment:
Instead of using version sniffing here, what do you think about using feature sniffing?  I've been generally sympathetic to the idea that keeping track of features (especially undocumented or private APIs) by versions leads to more complex code over time, and can sometimes bite you when/if a future version reverts back to an old prop name.  So instead of this version checking, we might do something like this 

```
const connOptsFromRedisClient = (rc) => {    
    if(rc.connection_options) {         
        return rc.connection_options
    }
    if(rc.connection_option) {          
        return rc.connection_option
    }
    if(rc.connectionOption) {
        return rc.connectionOption
    }
    if(rc.connection_options) {
        return rc.connection_options
    }       
    
    return {} 
}
```

Reply:
I can generally get behind biasing towards feature-sniffing over version-switching. Initially I'd used this approach to try to do the detection in the init, rather than having to feature-sniff for each call of the wrapped send command. However that is likely premature optimization.

I doubt node-redis will revert back to an old prop name. One (again unlikely) benefit of feature detection is if someone is playing with a personal build of node-redis that isn't one of the tagged releases.

Objections to shortening the above to:

```
const connOptsFromRedisClient = (rc) => rc.connection_options || rc.connection_option || rc.connectionOption || {}
```

?

Reply:
> premature optimization

fair and reasonable. 

> Objections to shortening the above to:

zero objections, have at it.